pkg/scheduler/queue/queue_test.go

```diff
@@ -18,102 +19,126 @@ import (
-			q.RegisterQuerierConnection(fmt.Sprintf("querier-%d", ix))
-		}
-
-		queues = append(queues, q)
-
-		for j := 0; j < numTenants; j++ {
-			requests = append(requests, fmt.Sprintf("%d-%d", n, j))
-			users = append(users, strconv.Itoa(j))
-		}
-	}
-
-	b.ResetTimer()
-	for n := 0; n < b.N; n++ {
-		for i := 0; i < maxOutstandingPerTenant; i++ {
-			for j := 0; j < numTenants; j++ {
-				err := queues[n].EnqueueRequest(users[j], requests[j], 0, nil)
-				if err != nil {
-					b.Fatal(err)
-				}
+func BenchmarkConcurrentQueueOperations(b *testing.B) {
+	req := "the query request"
+	maxQueriers := 0 // TODO: test with limit enforced?
+
+	for _, numTenants := range []int{1, 10, 1000} {
+		b.Run(fmt.Sprintf("%v tenants", numTenants), func(b *testing.B) {
+			for _, numProducers := range []int{10, 25} { // Query-frontends run 5 parallel streams per scheduler by default, and we typically see 2-5 frontends running at any one time.
+				b.Run(fmt.Sprintf("%v concurrent producers", numProducers), func(b *testing.B) {
+					for _, numConsumers := range []int{16, 160, 1600} { // Queriers run with parallelism of 8 when query sharding is enabled.
+						b.Run(fmt.Sprintf("%v concurrent consumers", numConsumers), func(b *testing.B) {
+							queueLength := promauto.With(nil).NewGaugeVec(prometheus.GaugeOpts{}, []string{"user"})
+							discardedRequests := promauto.With(nil).NewCounterVec(prometheus.CounterOpts{}, []string{"user"})
+							enqueueDuration := promauto.With(nil).NewHistogram(prometheus.HistogramOpts{})
+							queue := NewRequestQueue(100, 0, queueLength, discardedRequests, enqueueDuration)
+
+							start := make(chan struct{})
+							producersAndConsumers, ctx := errgroup.WithContext(context.Background())
+							require.NoError(b, queue.starting(ctx))
+							b.Cleanup(func() {
+								require.NoError(b, queue.stop(nil))
+							})
+
+							requestCount := func(total int, instanceCount int, instanceIdx int) int {
+								count := total / instanceCount
+								totalCountWithoutAdjustment := count * instanceCount
+
+								if totalCountWithoutAdjustment >= total {
+									return count
+								}
+
+								additionalRequests := total - totalCountWithoutAdjustment
+
+								if instanceIdx < additionalRequests {
+									// If we can't perfectly spread requests across all instances, assign remaining requests to the first instances.
+									return count + 1
+								}
+
+								return count
+							}
+
+							// TODO: limit the rate of production?
+							startProducer := func(producerIdx int) error {
+								requestCount := requestCount(b.N, numProducers, producerIdx)
+								tenantID := producerIdx % numTenants
+								<-start
+
+								for i := 0; i < requestCount; i++ {
+									for {
+										err := queue.EnqueueRequest(strconv.Itoa(tenantID), req, maxQueriers, func() {})
+										if err == nil {
+											break
+										}
+
+										// Keep retrying if we've hit the max queue length.
```

Comment:
Is this comment accurate? Here we return immediately, don't retry. I think it's better if we keep retrying instead of giving up

Reply:
We retry if we've hit the limit, but return immediately otherwise. I've made the comment clearer in b55bda5.

Reply:
you're right, I had misread the code